File: conversation/src/main/java/ru/pchurzin/conversation/Transactions.java
package ru.pchurzin.conversation;

import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

public class Transactions {
    public static void main(String[] args) {

    }
}

@Component
class SomeService3 {

    @Transactional
    public void method1() {
        // saves some data to database
    }

    public void method2() {
        method1();
    }

    @Transactional
    private void method3() {
        // saves some data to database
    }

    public void method4() {
        method3();
    }

}

@Component
@Transactional
class SomeService4 {

    public void method1() throws InterruptedException {
        Thread.sleep(1000); // throws InterruptedException
    }

}

File: conversation/src/main/kotlin/ru/pchurzin/conversation/Transactions.kt
package ru.pchurzin.conversation

import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class SomeService {

    @Transactional
    fun method1() {
        // saves some data to database
    }

    fun method2() {
        method1()
    }

    fun method3() {
        method4()
    }

    @Transactional
    private fun method4() {
        // saves some data to database
    }
}

// =============================================================================

@Component
@Transactional
class SomeService2 {
    fun method1() {
        // saves some data to database
        Thread.sleep(1000) // throws InterruptedException
    }
}

inline fun <reified T : Any> getBean(): T = TODO()